File: app_qltv/lib/BackEnd3/app/controllers/baoCaoTonKhoLoaiVang.js
var mysql = require('mysql');
var express = require('express');
var {db, db2} = require('../config/index_2');
var ApiError = require('../api-error');

exports.baoCaoTonKhoLoaiVang = async (req, res, next)=>{
    try{
        //Lấy theo từng tên loại vàng trước
        db.query(
            `
            SELECT DISTINCT nh.NHOM_TEN , nh.NHOMHANGID
            FROM nhom_hang nh
                INNER JOIN danh_muc_hang_hoa dmh on nh.NHOMHANGID = dmh.NHOMHANGID
                JOIN loai_hang lh ON lh.LOAIID = dmh.LOAIID
                JOIN ton_kho tk ON tk.HANGHOAID = dmh.HANGHOAID
            WHERE dmh.SU_DUNG="1" AND nh.SU_DUNG = 1 AND lh.SU_DUNG = 1 AND  tk.SL_TON = 1
            `,
            async (err, result1)=>{
                if(err){
                    return res.status(404).json({message: `Loi khi lấy thông tin danh sách hàng hóa theo từng loại vàng - ${err}`});
                }else{
                    //TÌm kiếm theo tên loại vàng
                    const TimKiemTheoNhom = (query, tenNhom) =>{
                        return new Promise((resolve, reject)=>{
                            db.query(query, (err, result)=>{
                                if(err){
                                    return res.status(404).json({message: `Loi khi thực hiện lấy thông tin loại vàng - ${tenNhom} - ERROR: ${err}`});
                                }
                                if(!result || result.length == 0){
                                    return res.status(404).json({message: `Loi khong tim thay ten loai vang ${tenNhom}.`});
                                }
                                resolve(result);
                                return result;
                            });
                        });
                    }

                    let MangKQ = [];
                    let tong_TLThuc = 0,
                        tong_TL_hot =0,
                        tong_TLvang=0,
                        tong_CongGoc=0,
                        tong_GiaCong=0,
                        tong_ThanhTien =0,
                        SoLuong=0;
                    //Lấy danh sách từng tên loại vàng
                    for(const item of result1) {
                        //console.log(item.NHOM_TEN);
                        let ketqua = await TimKiemTheoNhom(`
                            SELECT nh.NHOM_TEN, 
                                dmh.HANGHOAMA, 
                                dmh.HANG_HOA_TEN, 
                                lh.LOAI_TEN,
                                dmh.CAN_TONG, 
                                dmh.TL_HOT, 
                                (dmh.CAN_TONG - dmh.TL_HOT) AS TL_vang, 
                                dmh.CONG_GOC, 
                                dmh.GIA_CONG,
                                nh.DON_GIA_BAN AS DonGiaBan,
                                (nh.DON_GIA_BAN * 10 * ((dmh.CAN_TONG - dmh.TL_HOT) / 1000))+dmh.GIA_CONG AS ThanhTien 
                            FROM nhom_hang nh
                                INNER JOIN danh_muc_hang_hoa dmh ON nh.NHOMHANGID = dmh.NHOMHANGID
                                JOIN loai_hang lh ON lh.LOAIID = dmh.LOAIID
                                JOIN ton_kho tk ON tk.HANGHOAID = dmh.HANGHOAID
                            WHERE nh.NHOMHANGID = "${item.NHOMHANGID}" 
                                AND dmh.SU_DUNG = 1 
                                AND nh.SU_DUNG = 1 
                                AND lh.SU_DUNG = 1 
                                AND dmh.SO_LUONG > 0 
                                AND tk.SL_TON = 1
                            ORDER BY dmh.HANGHOAMA;    
                        `,item.NHOMHANGID);
                  
                        //Tính tổng
                        for(const item of ketqua){
                            nhomTen = item.NHOM_TEN;
                            tong_TLThuc += item.CAN_TONG;
                            tong_TL_hot +=item.TL_HOT;
                            tong_TLvang +=item.TL_vang;
                            tong_CongGoc +=item.CONG_GOC;
                            tong_GiaCong +=item.GIA_CONG;
                            tong_ThanhTien += item.ThanhTien;
                            SoLuong++;
                        }

                        var tinhTong = {
                            "NhomTen": nhomTen,
                            "SoLuong:": SoLuong,
                            "TongTL_Thuc":tong_TLThuc, 
                            "TongTL_hot": tong_TL_hot, 
                            "TongTL_Vang": tong_TLvang, 
                            "TongCongGoc": tong_CongGoc, 
                            "TongGiaCong":tong_GiaCong,
                            "TongThanhTien": tong_ThanhTien,
                        }

                        //Đặt lại số lượng về 0
                        SoLuong=0;
                        tong_TLThuc = 0;
                        tong_TL_hot =0;
                        tong_TLvang=0;
                        tong_CongGoc=0;
                        tong_GiaCong=0;
                        tong_ThanhTien =0;
                        SoLuong=0;

                        let ketQuaTungLoai = {
                            data: ketqua,
                            sumary: tinhTong
                        };

                        MangKQ.push(ketQuaTungLoai)
                    }
                    return res.status(200).json(MangKQ);
                }
        });
    }catch(err){
        return next(new ApiError(500, `Loi xuat hiện khi thực hiện truy vấn bao cáo hàng hóa theo loại vàng: ${err.message}`));
    }
};

exports.baoCaoTonKhoLoaiVang_TenLoaiVang = async (req, res, next)=>{
    try{
        const NHOM_TEN = req.params.NHOM_TEN;
        db.query(`
            SELECT nh.NHOM_TEN, dmh.HANGHOAMA, dmh.HANG_HOA_TEN, lh.LOAI_TEN,dmh.CAN_TONG, dmh.TL_HOT, (dmh.CAN_TONG -  dmh.TL_HOT) TL_vang, dmh.CONG_GOC, dmh.GIA_CONG 
            FROM nhom_hang nh
            INNER JOIN danh_muc_hang_hoa dmh on nh.NHOMHANGID = dmh.NHOMHANGID
            JOIN loai_hang lh ON lh.LOAIID = dmh.LOAIID
            WHERE nh.NHOM_TEN="${NHOM_TEN}" AND dmh.SU_DUNG="1"
        `,(err, result)=>{
            if(err){
                return res.status(400).json({message: `Loi khi thực hiện lấy thông tin loại vàng - ${NHOM_TEN} - ERROR: ${err}`});
            }
            if(!result || result.length == 0){
                return res.status(404).json({message: `Loi khong tim thay ten lọai vàng ${NHOM_TEN}.`});
            }

            let tong_TLThuc = 0,
                tong_TL_hot =0,
                tong_TLvang=0,
                tong_CongGoc=0,
                tong_GiaCong=0,
                SoLuong=0;

            for(const item of result){
                tong_TLThuc += item.CAN_TONG;
                tong_TL_hot +=item.TL_HOT;
                tong_TLvang +=item.TL_vang;
                tong_CongGoc +=item.CONG_GOC;
                tong_GiaCong +=item.GIA_CONG;
                SoLuong++;
            }

            return res.status(200).json({
                data:result,
                sumary:{
                    "Số lượng:": SoLuong,
                    "Tổng TL_Thực":tong_TLThuc, 
                    "Tổng TL_hột": tong_TL_hot, 
                    "Tổng TL_Vàng": tong_TLvang, 
                    "Tổng Công gốc": tong_CongGoc, 
                    "Tổng giá công":tong_GiaCong
                }
            });
        });

    }catch(err){
        return next(new ApiError(500, `Loi xuat hiện khi thực hiện truy vấn bao cáo hàng hóa theo loại vàng: ${err.message}`));
    }
};

File: app_qltv/lib/FrontEnd/model/danhmuc/BaoCaoTonKhoLoaiVang.dart
class LoaiVangTonKho {
  final String? nhomTen;
  final String? hangHoaMa;
  final String? hangHoaTen;
  final double? canTong;
  final double? tlHot;
  final double? tlVang;
  final double? congGoc;
  final double? giaCong;
  final double? donGiaBan;
  final double? thanhTien;

  // Constructor
  LoaiVangTonKho({
    required this.nhomTen,
    required this.hangHoaMa,
    required this.hangHoaTen,
    required this.canTong,
    required this.tlHot,
    required this.tlVang,
    required this.congGoc,
    required this.giaCong,
    required this.donGiaBan,
    required this.thanhTien,
  });

  // Copy constructor
  LoaiVangTonKho copyWith({
    String? nhomTen,
    String? hangHoaMa,
    String? hangHoaTen,
    double? canTong,
    double? tlHot,
    double? tlVang,
    double? congGoc,
    double? giaCong,
    double? donGiaBan,
    double? thanhTien,
  }) {
    return LoaiVangTonKho(
      nhomTen: nhomTen ?? this.nhomTen,
      hangHoaMa: hangHoaMa ?? this.hangHoaMa,
      hangHoaTen: hangHoaTen ?? this.hangHoaTen,
      canTong: canTong ?? this.canTong,
      tlHot: tlHot ?? this.tlHot,
      tlVang: tlVang ?? this.tlVang,
      congGoc: congGoc ?? this.congGoc,
      giaCong: giaCong ?? this.giaCong,
      donGiaBan: donGiaBan ?? this.donGiaBan,
      thanhTien: thanhTien ?? this.thanhTien,
    );
  }

  // Factory constructor to create an instance from a Map
  factory LoaiVangTonKho.fromMap(Map<String, dynamic> map) {
    return LoaiVangTonKho(
      nhomTen: map['NHOM_TEN'] ?? '',
      hangHoaMa: map['HANGHOAMA'] ?? '',
      hangHoaTen: map['HANG_HOA_TEN'] ?? '',
      canTong: (map['CAN_TONG'] as num?)?.toDouble() ?? 0.0,
      tlHot: (map['TL_HOT'] as num?)?.toDouble() ?? 0.0,
      tlVang: (map['TL_vang'] as num?)?.toDouble() ?? 0.0,
      congGoc: (map['CONG_GOC'] as num?)?.toDouble() ?? 0.0,
      giaCong: (map['GIA_CONG'] as num?)?.toDouble() ?? 0.0,
      donGiaBan: (map['DonGiaBan'] as num?)?.toDouble() ?? 0.0,
      thanhTien: (map['ThanhTien'] as num?)?.toDouble() ?? 0.0,
    );
  }

  // Method to convert an instance to a Map
  Map<String, dynamic> toMap() {
    return {
      'NHOM_TEN': nhomTen,
      'HANGHOAMA': hangHoaMa,
      'HANG_HOA_TEN': hangHoaTen,
      'CAN_TONG': canTong,
      'TL_HOT': tlHot,
      'TL_vang': tlVang,
      'CONG_GOC': congGoc,
      'GIA_CONG': giaCong,
      'DonGiaBan': donGiaBan,
      'ThanhTien': thanhTien,
    };
  }
}

class BaoCaoTonKhoLoaiVang {
  final List<LoaiVangTonKho> data;
  final String nhomTen;
  final int? soLuong;
  final double? tongTlThuc;
  final double? tongTlHot;
  final double? tongTlVang;
  final double? tongCongGoc;
  final double? tongGiaCong;
  final double? tongThanhTien;

  // Constructor
  BaoCaoTonKhoLoaiVang({
    required this.data,
    required this.nhomTen,
    required this.soLuong,
    required this.tongTlThuc,
    required this.tongTlHot,
    required this.tongTlVang,
    required this.tongCongGoc,
    required this.tongGiaCong,
    required this.tongThanhTien,
  });

  // Factory constructor to create an instance from a Map
  factory BaoCaoTonKhoLoaiVang.fromMap(Map<String, dynamic> json) {
    var list = json['data'] as List;
    List<LoaiVangTonKho> dataList =
        list.map((i) => LoaiVangTonKho.fromMap(i)).toList();

    return BaoCaoTonKhoLoaiVang(
      data: dataList,
      nhomTen: json['sumary']['NhomTen'] ?? '',
      soLuong: json['sumary']['SoLuong'] ?? 0,
      tongTlThuc: (json['sumary']['TongTL_Thuc'] as num?)?.toDouble() ?? 0.0,
      tongTlHot: (json['sumary']['TongTL_hot'] as num?)?.toDouble() ?? 0.0,
      tongTlVang: (json['sumary']['TongTL_Vang'] as num?)?.toDouble() ?? 0.0,
      tongCongGoc: (json['sumary']['TongCongGoc'] as num?)?.toDouble() ?? 0.0,
      tongGiaCong: (json['sumary']['TongGiaCong'] as num?)?.toDouble() ?? 0.0,
      tongThanhTien:
          (json['sumary']['TongThanhTien'] as num?)?.toDouble() ?? 0.0,
    );
  }
}
